runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -2937,19 +2937,259 @@ J9::Z::TreeEvaluator::anewArrayEvaluator(TR::Node * node, TR::CodeGenerator * cg
+
+   // Branch to OOL if there's not enough space for an array of size 0.
+   TR::Register *temp1Reg = cg->allocateRegister();
+   generateRIEInstruction(cg, TR::InstOpCode::getAddHalfWordImmDistinctOperandOpCode(), node, temp1Reg, targetReg, zeroArraySize);
+   generateRXInstruction(cg, TR::InstOpCode::getCmpLogicalOpCode(), node, temp1Reg, generateS390MemoryReference(vmThreadReg, offsetof(J9VMThread, heapTop), cg));
+   generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BH, node, oolJumpLabel);
+
+   // If there's enough space, then we can continue to allocate.
+   generateRXInstruction(cg, TR::InstOpCode::getStoreOpCode(), node, temp1Reg, generateS390MemoryReference(vmThreadReg, offsetof(J9VMThread, heapAlloc), cg));
+   
+   bool use64BitClasses = comp->target().is64Bit() && !TR::Compiler->om.generateCompressedObjectHeaders();
+
+   // Init class and fields, then jump to end of ICF
+   generateRXInstruction(cg, use64BitClasses ? TR::InstOpCode::STG : TR::InstOpCode::ST, node, classReg, generateS390MemoryReference(targetReg, TR::Compiler->om.offsetOfObjectVftField(), cg));
+   generateSILInstruction(cg, use64BitClasses ? TR::InstOpCode::MVGHI : TR::InstOpCode::MVHI, node, generateS390MemoryReference(targetReg, fej9->getOffsetOfContiguousArraySizeField(), cg), 0);
+   generateSILInstruction(cg, use64BitClasses ? TR::InstOpCode::MVGHI : TR::InstOpCode::MVHI, node, generateS390MemoryReference(targetReg, fej9->getOffsetOfDiscontiguousArraySizeField(), cg), 0);
+   generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BRC, node, cFlowRegionEnd);
+
+   // We end up in this region of the ICF if the first dimension is non-zero and the second dimension is zero.
+   generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, nonZeroFirstDimLabel);
+
+   TR::Register *componentClassReg = cg->allocateRegister();
+   generateRXInstruction(cg, TR::InstOpCode::getLoadOpCode(), node, componentClassReg, generateS390MemoryReference(classReg, offsetof(J9ArrayClass, componentType), cg));
+
+   // Calculate maximum allowable object size in elements and jump to OOL if it's higher than firstDimLenReg.
+   int32_t elementSize = TR::Compiler->om.sizeofReferenceField();
+   uintptr_t maxObjectSize = cg->getMaxObjectSizeGuaranteedNotToOverflow();
+   uintptr_t maxObjectSizeInElements = maxObjectSize / elementSize;
+
+   if (comp->target().is64Bit() && !(maxObjectSizeInElements >= 0 && maxObjectSizeInElements <= (uintptr_t)INT_MAX))
+      {
+      generateS390CompareAndBranchInstruction(cg, TR::InstOpCode::CLG, node, firstDimLenReg, static_cast<int64_t>(maxObjectSizeInElements), TR::InstOpCode::COND_BRH, oolJumpLabel, false);
+      }
+   else
+      {
+      generateS390CompareAndBranchInstruction(cg, TR::InstOpCode::CL, node, firstDimLenReg, static_cast<int32_t>(maxObjectSizeInElements), TR::InstOpCode::COND_BHR, oolJumpLabel, false);
+      }
```

Comment:
This is incorrect. `firstDimLenReg` is a 32-bit value, so it would be functionally incorrect to test this value against a 64-bit constant. The upper 32-bits are dirty. IMO we should just have a check up front where we check `!secondChild->getOpCode().isLoadConst()` and just ensure the array is not greater than `cg->getMaxObjectSizeGuaranteedNotToOverflow()`, and if it is just bail out of the inline sequence.

We're only trying to accelerate the most common cases here. If someone is allocating a 2D array of size > 0x10000000 we should just let the helper handle it.

Reply:
Do you mean to check the size of the array at compile time and bail if it's too large? The second child is not the size of the array. It represents the numbers of dimensions in the multi-dimensional array. So I don't think we can do the check you proposed above as that information is not available to us at compile time.

Reply:
As per our offline discussion, we can't do the check proposed during compile time. However we should only ever do a 32-bit comparison here. I've simplified this check in 6678b56.